File: 237/delete-node-in-a-linked-list.py
# 這個題目比較麻煩，因若第1個node被刪除的話，不知道要怎麼刪、刪的話就沒有head了
# "You will not be given access to the first node of head."
# 這樣就沒辦法由 head 繞過「要刪除的node」
# 不過有個巧妙的解法，是 copy 下一筆的值，再「繞過」下一筆，就可以了
# 那最後一筆怎麼辦？ 測資 [4,5,1,9] 9 告訴我們 "cannot delete the tail."
# 所以太好了，不用管這個狀況。
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteNode(self, node):
        """
        :type node: ListNode
        :rtype: void Do not return anything, modify node in-place instead.
        """
        next = node.next
        node.val = next.val # 把下一筆的值 copy 過來
        node.next = next.next # 把下一筆的 next 指標 copy 過來。
        # 也就是「要刪除的值」不見了，變身下一筆。
        # 自己當成「下一筆」的替身，再殺「下一筆」的本尊

File: 237/delete-node-in-a-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
// 今天 LeetCode 挑戰題，可以說很難
//（Median中等題，要用到大二資料結構的Linked List觀念，又需要拐個奇怪的彎），
// 也可以說很簡單（程式碼只需要2行）。
// 題目是給你一個指標，指到一個 node 資料結構。然後想把它刪掉。
// 但是，你又不能「真的把它刪掉」，因為你「只有這個node」及之後的nodes的資訊，
// 而沒有「前一個node」的資訊，所以真的無法「刪不掉它」。
// 解法：拐個彎，不需要把它刪掉，而是把它「打扮成下一個node」，
//      再把「下一個node犧牲忽略掉」就完成了。
void deleteNode(struct ListNode* node) {
    // 題目保證不是最後一點，所以「存在node->next」
    node->val = node->next->val;  // 把下一筆的「值」往前移
    node->next = node->next->next;  // 把下一筆的「指標」往前移    
}
